fix: dot and empty paths normalized correctly for COPY to WORKDIR

profnandaa opened the pull request:
In my previous fix on #4825, I had removed this line knowing that all that had been addressed in `pahtRelativeToWorkingDir`:

```go
	if cfg.params.DestPath == "." // <-- this one
		|| cfg.params.DestPath == ""
                || cfg.params.DestPath[len(cfg.params.DestPath)-1] == filepath.Separator {
		dest += string(filepath.Separator)
	}
```

However, I had overlooked the `"."` and `""` scenarios. `""`. The  `"/"`case is handled correctly in `system.NormalizePath()`.

This change therefore undoes this, to make sure "." is transformed correctly to "./" during COPY operation, same to "" -> "./". This is important especially for WORKDIR that are not `/`, so that `COPY --link` operations are handled properly.

fixes #5070

profnandaa commented:
While were here, I was torn between adding this case to `NormalizePath` or reverting to how it was initially.
Also wanted to understand the intention for this block: @gabriel-samfira 

https://github.com/moby/buildkit/blob/997156aacc32eb2856256e5088e1bd927484c35d/util/system/path.go#L77-L82

tonistiigi reviewed (commented):
We need regression integration test for this as well. Is this something that should be handled by `pathRelativeToWorkingDir` (that part could be follow up).

profnandaa commented on frontend/dockerfile/dockerfile2llb/convert.go:
```diff
@@ -1763,9 +1763,6 @@ func pathRelativeToWorkingDir(s llb.State, p string, platform ocispecs.Platform)
 		return "", err
 	}
 
-	if len(p) == 0 {
```
nit: removed so that empty directory `""` are handled properly, especially for non-root `WORKDIR`. That's where the regression happens.

profnandaa commented:
> We need regression integration test for this as well. Is this something that should be handled by `pathRelativeToWorkingDir` (that part could be follow up).

Sure, it's better at `pathRelativeToWorkingDir`, I've moved it and also fixed how "" dir's were being handled.

Added the integration tests. It was very interesting that the tests were passing even without the fix, if you have `/` as the WORKDIR; and was getting a little thrown off. However, fixed with: `WORKDIR /var/www` and could repro the regression correctly.

profnandaa commented:
@tonistiigi - PTAL again.
